APR Services - Add some yield tokens APR to linear pools

lgahdl opened the pull request:
Adding APR Service to fetch token yields APR on https://yield-tokens.balancer.workers.dev/ and calculate the APR for boosted pools with nested pools that contain the fetched tokens.

Commit:
Created a service to fetch yield tokens APR and push into DB

Commit:
Merge remote-tracking branch 'origin/v3-canary' into non-reaper-yield-tokens-apr

Commit:
Adding Apr Service to Mainnet network

gmbronco commented:
What are "Non Reaper Boosted pools"?

lgahdl commented:
> What are "Non Reaper Boosted pools"?

I thought it would be necessary to update boosted pools in the same APR Service that updates the linear pools with "non reaper tokens" yield (Which John called Non Reaper Boosted Pools, because there's another service that updates the APR for pools with wrapped tokens from Reaper Crypt) but there's already a service that updates all boosted pools, so it's not necessary to update the "Non Reaper Boosted Pools", because they're already being updated by the "boosted-pool.apr.service"

Commit:
Added yield tokens fetch funcionality directly on beethovenx-backend, instead of use the yield-tokens cloudflare service;
Removed the boosted pools apr, and let only the yield tokens APR update, boosted pools are updated in another service;

franzns commented on modules/pool/lib/apr-data-sources/ib-tokens-apr.service.ts:
```diff
@@ -0,0 +1,52 @@
+              title: `${ token.token.symbol || token.address } APR`,
+              apr: aprs.get(token.address) ?? 0,
+              group: null,
+              type: 'IB_YIELD',
```
This needs to be different, based on the pool. If it's a wrapped asset from a linear pool, it needs to be `LINEAR_BOOSTED`, else `IB_YIELD`

franzns commented on modules/pool/lib/apr-data-sources/ib-tokens-apr.service.ts:
```diff
@@ -0,0 +1,52 @@
+    await prismaBulkExecuteOperations(operations);
+  }
+
+  private async fetchYieldTokensApr(): Promise<Map<string, number>> {
+    const res = await axios.get("https://yield-tokens.balancer.workers.dev/");
+    const aprs = new Map<string, number>(Object.entries(res.data as {
+      [key: string]: number
+    }).map(([key, apr]) => [key, apr * 0.0001]));
+    return aprs;
+  }
```
I assume this will change to be sourced locally from the sources you added?

Commit:
Adapting contracts from apr handler to ethers instead of viem;

Commit:
Changing apr type if pool type is linear;

Commit:
removing viem

Commit:
Merge commit 'fccc29b6e65ffd3df41b7b7de0c2e407e8de2c59' into non-reaper-yield-tokens-apr
Grouping APRs by source;

Commit:
Fixing Ovix APR;

franzns commented on modules/token/lib/token-apr-handler/fetch-all.ts:
```diff
@@ -0,0 +1,16 @@
+import { tokens } from "./tokens";
+import { TokenApr } from "./types";
+
+export const fetchAllAprs = async (): Promise<TokenApr[]> => {
```
Please create a proper function for this

franzns commented on modules/token/lib/token-apr-handler/sources/reaper.ts:
```diff
@@ -0,0 +1,50 @@
+
+
+const jsonRpcProvider = new JsonRpcProvider('https://arb1.arbitrum.io/rpc', 42161 /*ARBITRUM*/)
+const provider = MulticallWrapper.wrap(jsonRpcProvider)
```
why is this multicall wrapper needed?

lgahdl replied:
I used it so calls performed simultaneously can be automatically batched. Like in this case:
```
const getAprs = () => {

  const calls = Object.keys(strategiesMap).map(async (coin) => {
    const contract = new Contract(strategiesMap[coin], abi, provider)
    return contract.averageAPRAcrossLastNHarvests(3)
  })

  return Promise.all(calls)
}
```

on viem, it was this:
```
const getAprs = () => {
  const contracts = Object.keys(strategiesMap).map((coin) => ({
    address: strategiesMap[coin],
    abi,
    functionName: 'averageAPRAcrossLastNHarvests',
    args: [3],
  }))

  return client.multicall({ contracts })
}

```

franzns commented on modules/token/lib/token-apr-handler/fetch-all.ts:
```diff
@@ -0,0 +1,16 @@
+import { TokenApr } from "./types";
+
+export const fetchAllAprs = async (): Promise<TokenApr[]> => {
+  const res = await Promise.all(
```
any of these fetches can fail for various reasons. I'd rather do that sequential (no need to optimize in the cron) and log proper errors to sentry

franzns commented on modules/token/lib/token-apr-handler/fetch-all.ts:
```diff
@@ -0,0 +1,16 @@
+    tokens.map(async ({ name, group, fetchFn }) => {
+      const fetchedResponse: {[key:string]:number} = await fetchFn()
+      return Object.entries(fetchedResponse).map(([address, aprValue]) => ({
+        val: aprValue * 0.0001 /*Values come in BPS 10000=>100% */,
```
maybe the apr services should already provide values in %

franzns commented:
Looked through the thing now. We will need to redesign it a bit and make it multichain aware. Currently, some APR sources are multichain aware and others arent. 
I would even redesign the whole thing, introducing interfaces for the APR sources, similar to what we did for the other APR services. 
In addition, that thing should live in the pools module with the other APR code and not in the token module.

Commit:
Redesigning Ib Apr Handlers to be more scalable and open to new networks, tokens and urls;

Commit:
adding interface to default, ankr and aave apr handlers;

franzns commented on modules/pool/lib/apr-data-sources/ib-yield-apr-handlers/ib-yield-apr-handlers.ts:
```diff
@@ -0,0 +1,28 @@
+    this.handlers = handlers;
+  }
+  
+  getHandlersAprs = async (): Promise<TokenApr[]> => {
+    const aprPromises = this.handlers.map(async (handler) => {
+      const fetchedResponse: { [key: string]: number } = await handler.getAprs()
+      return Object.entries(fetchedResponse).map(([address, aprValue]) => ({
+        val: aprValue,
+        group: handler.group,
+        address
+      }))
+    });
+    const res = Array(this.handlers.length)
+    for (const [index, aprPromise] of aprPromises.entries()) {
+      res[index] = await aprPromise
+    }
+    return res.flat();
+  }
+}
```
Make this a proper function of this class

franzns commented on modules/pool/lib/apr-data-sources/ib-yield-apr-handlers/sources/index.ts:
```diff
@@ -0,0 +1,50 @@
+import { tetuAprHandler } from "./tetu/tetu-apr-handler";
+import { tranchessMainnetAprHandler } from "./tranchess/tranchess-apr-handler";
+
+export const aprHandlers: AprHandler[] = [
+  aaveV2MainnetAprHandler,
+  aaveV2PolygonAprHandler,
+  aaveV3MainnetAprHandler,
+  aaveV3PolygonAprHandler,
+  aaveV3ArbitrumAprHandler,
+  ankrEthMainnetAprHandler,
+  stMaticAprHandler,
+  cbEthAprHandler,
+  sfrxEthAprHandler,
+  rETHAprHandler,
+  USDRAprHandler,
+  MATICXAprHandler,
+  wbETHAprHandler,
+  swETHAprHandler,
+  wjAURAAprHandler,
+  eulerMainnetAprHandler,
+  gearboxMainnetAprHandler,
+  idleMainnetAprHandler,
+  overnightMainnetAprHandler,
+  ovixZkEVMAprHandler,
+  reaperArbitrumAprHandler,
+  tesseraApePoolAprHandler,
+  tetuAprHandler,
+  tranchessMainnetAprHandler,
+]
```
Should these better be initialized in the ib-yield-apr-handlers.ts instead of an additional file?

franzns commented on modules/pool/lib/apr-data-sources/ib-yield-apr-handlers/types.ts:
```diff
@@ -0,0 +1,10 @@
+}
+
+export interface AprHandler {
+  readonly group: string | undefined;
```
Should this ever be undefined?

Commit:
Adding multichain support;
Changing arrow functions to normal functions;

Commit:
Types files for ankr and default-fetch; (Missing in the last commit)

Commit:
Merge commit 'a0eb08ebd453441e8f566fb03cc684ec8b4e324d' into non-reaper-yield-tokens-apr

# Conflicts:
#	yarn.lock

Commit:
Creating new migration file;

Commit:
Removing test route;
Adding migration file;

Commit:
removing unused imports;

Commit:
Removing ethers-multicall-provider package

Commit:
removing undefined from "group" type

franzns commented on modules/pool/lib/apr-data-sources/ib-tokens-apr.service.ts:
```diff
@@ -0,0 +1,68 @@
+              title: `${ tokenSymbol } APR`,
+              apr: aprs.get(token.address)?.val ?? 0,
+              group: (aprs.get(token.address)?.group as PrismaPoolAprItemGroup) ?? null,
+              type: pool.type === 'LINEAR' ? 'LINEAR_BOOSTED' : 'IB_YIELD',
```
that is not enough, you'll need to check if it is the wrapped token when it is in a linear pool. only then it is linear_boosted, otherwise its IB_yield. E.g. for a wsteth/rfWSTETH linear pool

franzns commented on modules/pool/lib/apr-data-sources/ib-yield-apr-handlers/sources/index.ts:
```diff
@@ -0,0 +1,25 @@
+import { aaveHandlers } from "./aave/aave-apr-handler";
+import { ankrHandlers } from "./ankr/ankr-apr-handler";
+import { defaultHandlers } from "./default-fetch/default-apr-handler";
+import { eulerHandlers } from "./euler/euler-apr-handler";
+import { gearboxHandlers } from "./gearbox/gearbox-apr-handler";
+import { idleAprHandlers } from "./idle/idle-apr-handler";
+import { overnightHandlers } from "./overnight/overnight-apr-handler";
+import { ovixHandlers } from "./ovix/ovix-apr-handler";
+import { reaperHandlers } from "./reaper/reaper-apr-handler";
+import { tesseraHandlers } from "./tessera/tessera-apr-handler";
+import { tetuHandlers } from "./tetu/tetu-apr-handler";
+
+export const aprHandlers = [
+  ...aaveHandlers,
+  ...ankrHandlers,
+  ...defaultHandlers,
+  ...eulerHandlers,
+  ...gearboxHandlers,
+  ...idleAprHandlers,
+  ...overnightHandlers,
+  ...ovixHandlers,
+  ...reaperHandlers,
+  ...tesseraHandlers,
+  ...tetuHandlers,
+]
```
include this into the ib-yield-apr-handlers.ts file, no need to do it separately

franzns commented on modules/pool/lib/apr-data-sources/ib-yield-apr-handlers/ib-yield-apr-handlers.ts:
```diff
@@ -0,0 +1,26 @@
+    this.handlers = aprHandlers.filter((handler) => handler.network === network);
+  }
+  
+   async getHandlersAprs (): Promise<TokenApr[]>  {
+    const aprPromises = this.handlers.map(async (handler) => {
+      const fetchedResponse: { [key: string]: number } = await handler.getAprs()
+      return Object.entries(fetchedResponse).map(([address, aprValue]) => ({
+        val: aprValue,
+        group: handler.group,
+        address
+      }))
+    });
+    const res = Array(this.handlers.length)
+    for (const [index, aprPromise] of aprPromises.entries()) {
+      res[index] = await aprPromise
+    }
+    return res.flat();
+  }
+}
```
couldn't we do this is a simple for loop? this is super hard to read

franzns commented on modules/pool/lib/apr-data-sources/ib-yield-apr-handlers/ib-yield-apr-handlers.ts:
```diff
@@ -0,0 +1,26 @@
+    this.handlers = aprHandlers.filter((handler) => handler.network === network);
+  }
+  
+   async getHandlersAprs (): Promise<TokenApr[]>  {
```
please rename to something like "fetchAprsFromAllHandlers"

franzns commented on modules/pool/lib/apr-data-sources/ib-yield-apr-handlers/types.ts:
```diff
@@ -0,0 +1,10 @@
+export type TokenApr = {
+  val: number;
+  address: string;
+  group?: string;
+}
+
+export interface AprHandler {
+  readonly group: string;
+  getAprs(): Promise<{ [key: string]: number }>;
+}
```
could these types live in the ib-yield-apr-handlers.ts without creating circular dependencies?

franzns commented on modules/pool/lib/apr-data-sources/ib-yield-apr-handlers/types.ts:
```diff
@@ -0,0 +1,10 @@
+
+export interface AprHandler {
+  readonly group: string;
+  getAprs(): Promise<{ [key: string]: number }>;
```
key is actually token address, right? rename it if so

franzns commented on modules/network/arbitrum.ts:
```diff
@@ -160,6 +162,7 @@ export const arbitrumNetworkConfig: NetworkConfig = {
     poolAprServices: [
+        new IbTokensAprService(arbitrumNetworkData.chain.id),
         new WstethAprService(tokenService, arbitrumNetworkData.lido!.wstEthContract),
         new ReaperCryptAprService(
```
we should be able to remove these if we have the IbTokensAprService

Commit:
Moving tokens and types to apr-handler file;
Changing map to for..of loop

Commit:
Using prismaId instead of chain id;
moving apr handler types file to ib-yield-apr-handlers;

timjrobinson commented on modules/pool/lib/apr-data-sources/ib-yield-apr-handlers/sources/aave-apr-handler.ts:
```diff
@@ -0,0 +1,232 @@
+    }
+}
+
+const wrappedAaveTokensV2Mainnet = {
```
I think all of this config makes more sense in the network config files. Or something like:

```
aaveTokens: {
  [Network.Mainnet]: {
    tokensV2: {
      ...
    },
    wrappedTokensV2: {
      ...
    }, 
    tokensV3: {
      ...
    }
 }
}
```

(The underlying tokens should be defined somewhere else as they are not Aave specific)

Then having one function `getAaveAprHander(network: Network, version: number)` that pulls the correct config settings and returns a new `AaveAprHandler` with it, instead of defining a function for every network. 

In general network names should never need to be added to variable names, it's better to use a map. 

This will make it much easier to extend as we add more networks, as eventually we may have 50+ networks and this will become messy.

timjrobinson commented on modules/pool/lib/apr-data-sources/ib-yield-apr-handlers/sources/default-apr-handler.ts:
```diff
@@ -0,0 +1,183 @@
+    path?: string;
+};
+
+const vETHMainnet = '0x4bc3263eb5bb2ef7ad9ab6fb68be80e43b43801f';
```
Like Aave, This could all be refactored into a map and all the handlers could be one function where you just pass in the network and it generates all handlers for that network on the fly and returns them in an array. Or you could pass in Network + Token and it returns a handler for that token if there is one defined.

Commit:
Checking if the pool is linear boosted by the token address;
Fixing network id type error on json rpc provider creation;

Commit:
Renaming ib-yield-apr-handlers to base-apr-handlers
Moving tokens to base-apr-handlers:
- Ankr eth fantom, ankr ftm fantom;
- Overnight Optimism usd+ and dai+ (moved overnight aprs to default base-apr-handlers)
- wstETH Optimism

Didn't move:
- Stader-staked-ftm <- Fixed Base APR Value?
- spooky <- Not being used on Fantom, so I didn’t add it to base apr handlers
- Yearn <- Using Base APR’s based on wrapped tokens in linear pools, not able to define which tokens to take
- Rocket Pool staked <- Fixed rETH base APR Value, I have rETH APR from url on main net, maybe I can use the same? https://drop-api.stafi.io/reth/v1/poolData
- Beefy <- Using beefy linear pools Id’s to get vaultAddress, there’s 2 options here: 1 - Let the beefy apr handler separated 2 - Create APR Handlers for each vault/beefy token.

- Reaper-Crypt <- The current apr handler verifies if the pool is from a ERC 4626 Factory, and build contract from wrappedToken.address, So I didn’t move this reaper to the base apr handler, same problem as Beefy

Commit:
Removing ib-yield-apr-handlers folder

Commit:
Refactoring all code to be more scalable;
Moving the apr config to network config;

Commit:
removing aave tokens unused;

timjrobinson commented on modules/pool/lib/apr-data-sources/base-apr-handlers/base-apr-handlers.ts:
```diff
@@ -0,0 +1,172 @@
+    group?: string;
+};
+
+// export const wrappedBoostedTokens = [
```
comments to delete

timjrobinson reviewed (commented):
Nice work, this looks much cleaner!

franzns commented on modules/network/mainnet.ts:
```diff
@@ -115,6 +116,201 @@ const mainnetNetworkData: NetworkData = {
+        idle: {
+            sourceUrl: 'https://api.idle.finance/junior-rates/',
+            authorizationHeader:
+                'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjbGllbnRJZCI6IkFwcDciLCJpYXQiOjE2NzAyMzc1Mjd9.L12KJEt8fW1Cvy3o7Nl4OJ2wtEjzlObaAYJ9aC_CY6M',
```
Does this need to be a secret?

lgahdl replied:
To get this bearer token, all it needs is fill a typeform, I think it's not a sensitive datum

franzns commented on modules/pool/lib/apr-data-sources/base-apr-handlers/sources/tessera-apr-handler.ts:
```diff
@@ -0,0 +1,60 @@
+    readonly group = 'TESSERA';
+
+    constructor(aprHandlerConfig: TesseraAprHandlerConfig) {
+        this.networkPrismaId = aprHandlerConfig.networkPrismaId;
```
Is this needed?

lgahdl replied:
Well noted! No, is not needed, will remove it

franzns commented on modules/pool/lib/apr-data-sources/ib-tokens-apr.service.ts:
```diff
@@ -0,0 +1,85 @@
+        });
+        for (const pool of tokenYieldPools) {
+            const totalLiquidity = pool.dynamicData?.totalLiquidity;
+            for (const token of pool.tokens) {
```
In theory, we should check whether the token has a rate provider set, as we only have an APR if there is a rate provider. We don't store that information in the backend though. Should probably add a TODO

franzns commented on modules/pool/lib/apr-data-sources/base-apr-handlers/base-apr-handlers.ts:
```diff
@@ -0,0 +1,172 @@
+    async fetchAprsFromAllHandlers(): Promise<TokenApr[]> {
+        let aprs: { val: number; group: string; address: string }[] = [];
+        for (const handler of this.handlers) {
+            const fetchedResponse: { [key: string]: number } = await handler.getAprs();
```
Add try/catch here to make sure a faulty apr handler doesnt block all other services. Log to sentry with APR handlers name/ID

franzns commented on modules/pool/lib/apr-data-sources/base-apr-handlers/sources/tetu-apr-handler.ts:
```diff
@@ -0,0 +1,40 @@
+                .map((t) => [t.vault, t.apr / 100]);
+            return Object.fromEntries(aprs);
+        } catch (error) {
+            console.error('Failed to fetch Tetu APR:', error);
```
These should also log to sentry with some kind of ID of this handler. Seem we need to introduce a handler ID as well.

franzns replied:
Above is for all APR handlders, not only this one.

franzns commented on modules/pool/lib/apr-data-sources/base-apr-handlers/sources/reaper-apr-handler.ts:
```diff
@@ -0,0 +1,73 @@
+
+import { AprHandler } from '../base-apr-handlers';
+
+class ReaperAprHandler implements AprHandler {
```
This does not implement the multistrategy APR fetching which needs a subgraph. I suggest to adapt the reaperHandlderConfig for this

lgahdl replied:
I think we could remove this Reaper Apr Handler and let the Reaper Service separated as it is, add the Reaper to the new service would move the Reaper rules (sftmx, multi and single strategy) that already exists into a new service and add complexity to the new service, what do you think?

franzns replied:
I would like to have the "old" reaper handler integrated into the new service you are creating. Adapt the reaperHandlerConfig to take singlestrategy vault addresses and multi-strategy vault address so you don't need to "hack" the way I did it.

franzns commented on modules/pool/lib/apr-data-sources/base-apr-handlers/sources/reaper-apr-handler.ts:
```diff
@@ -0,0 +1,73 @@
+                return [this.yieldTokens[tokenName], callsAprResults[i].toNumber() / 1e4];
+            });
+
+            return Object.fromEntries(aprs);
```
I'd personally do this in one for-loop instead of multiple loops with `map` and `for`. I find this hard to read with all these `Object.` calls

franzns commented on modules/network/network-config-types.ts:
```diff
@@ -100,6 +100,7 @@ export interface NetworkData {
         address: string;
         excludedFarmIds: string[];
     };
+    aprConfig: AprConfig;
```
rename this to baseIbAprConfig or similar. Needs to reflect this is only for the base APR config for any IB tokens

franzns commented on modules/pool/lib/apr-data-sources/apr-types.ts:
```diff
@@ -1,3 +1,5 @@
+import { Dictionary } from 'lodash'
+
```
remove unused

franzns replied:
still there :)

franzns commented on modules/pool/lib/apr-data-sources/liquid-staked-base-apr.service.ts:
```diff
@@ -1,6 +1,7 @@
 import axios from 'axios';
```
this whole class should be removed and integrated into the base-apr-handlers

franzns commented on modules/pool/lib/apr-data-sources/base-apr-handlers/base-apr-handlers.ts:
```diff
@@ -0,0 +1,172 @@
+import { TranchessAprHandler } from './sources/tranchess-apr-handler';
+import { AprConfig } from '../../../../network/network-config-types';
+
+export class BaseAprHandlers {
```
rename this to reflect this is only for IB and Linear APRs

franzns commented on modules/pool/lib/apr-data-sources/base-apr-handlers/sources/aave-apr-handler.ts:
```diff
@@ -0,0 +1,98 @@
+        };
+    };
+    subgraphUrl: string;
+    networkPrismaId: string;
```
For all handlers: I think these prismaId as well as networkId are not used anymore and can be removed.

franzns commented on modules/pool/lib/apr-data-sources/base-apr-handlers/sources/default-apr-handler.ts:
```diff
@@ -0,0 +1,62 @@
+    path: string;
+    scale: number;
+    networkPrismaId: string;
+    group = 'DEFAULT';
```
this should be empty or undefined and only set if needed, otherwise if a pool as reth and wsteth apr it would be grouped under DEFAULT in the UI.

franzns commented on modules/pool/lib/apr-data-sources/base-apr-handlers/sources/default-apr-handler.ts:
```diff
@@ -0,0 +1,62 @@
+        this.networkPrismaId = aprHandlerConfig.networkPrismaId;
+        this.path = aprHandlerConfig.path ?? '';
+        this.scale = aprHandlerConfig.scale ?? 100;
+        this.group = aprHandlerConfig.group ?? 'DEFAULT';
```
this should be empty or undefined and only set if needed, otherwise if a pool as reth and wsteth apr it would be grouped under DEFAULT in the UI.

franzns reviewed (commented):
Thank you Luiz and Tim. This looks so much better! I have added comments where I found issues. Some comments I only added to one handler and wrote if they affect all of them.

In general, I like this approach way more! Go ahead and migrate all other IB APR services like this.

Commit:
Renaming aprConfig to IbAprConfig;
Adding ReaperCrypt to Apr Handlers;
Removing services already included in base-apr-handlers;

Commit:
Removing Reaper Crypt Apr Service;

Commit:
Removing services that are already in the ib-yield service;
Adding fixed apr tokens;
Removing logs

Commit:
Removing test route;

Commit:
Merge branch 'v3-canary' into non-reaper-yield-tokens-apr

# Conflicts:
#	modules/network/arbitrum.ts
#	modules/network/avalanche.ts
#	modules/network/fantom.ts
#	modules/network/gnosis.ts
#	modules/network/mainnet.ts
#	modules/network/network-config-types.ts
#	modules/network/optimism.ts
#	modules/network/polygon.ts
#	modules/network/zkevm.ts
#	modules/pool/lib/apr-data-sources/fantom/yearn-vault-apr.service.ts
#	modules/pool/lib/apr-data-sources/reaper-crypt-apr.service.ts

Commit:
fixing base network incompatibilities;

Commit:
fixing optional beefy property in network config;

Commit:
Removing all unused functions of liquid staked base apr, left just xBoo, because idk what to do with the spooky apr service yet;

Commit:
Renaming AprConfig to IbAprConfig;

franzns commented on modules/pool/lib/apr-data-sources/ib-tokens-apr.service.ts:
```diff
@@ -0,0 +1,140 @@
+                // If pool is Linear
+                if (!isLinear) {
+                    continue;
+                }
```
not needed

franzns commented on modules/pool/lib/apr-data-sources/ib-tokens-apr.service.ts:
```diff
@@ -0,0 +1,140 @@
+                continue;
+            }
+            const isLinear = pool.type === 'LINEAR' && pool.linearData !== null;
+            if (!isLinear) {
```
this is essentially to find out whether its `IB_YIELD` or `LINEAR_BOOSTED`. That should be able to do it easier without duplicating all this code

franzns commented on modules/pool/lib/apr-data-sources/ib-linear-apr-handlers/sources/reaper-crypt-apr-handler.ts:
```diff
@@ -0,0 +1,141 @@
+            if (!token) {
+                return acc;
+            }
+            let tokenApr = parseFloat(apr) / APR_PERCENT_DIVISOR;
+            if (token.isSftmX) {
+                tokenApr = tokenApr * (1 + sFTMxBaseApr);
+            }
+            if (token.isWstETH) {
+                tokenApr = tokenApr * (1 + this.wstETHBaseApr);
+            }
```
is there a possibility to make this more general? Like if any of the IB-token-apr handlers for linear pools have an IB token, this needs to apply.

franzns commented on modules/pool/lib/apr-data-sources/reaper-crypt-apr.service.ts:
```diff
@@ -1,181 +0,0 @@
-import { isSameAddress } from '@balancer-labs/sdk';
```
remove alltogether

franzns commented on modules/pool/lib/apr-data-sources/fantom/yearn-vault-apr.service.ts:
```diff
@@ -1,59 +0,0 @@
-import { PoolAprService } from '../../../pool-types';
```
remove file

franzns reviewed (commented):
Thanks, still a few things. Also, I think Optimism is missing reaper single strat configs (maybe also fantom).

Also, there are two reaper apr handlers, I believe only one is needed.

franzns committed:
clean up types

franzns committed:
simplify ib-tokens-apr service

franzns committed:
change default handler, adapt token lists

franzns committed:
fix percentage

franzns committed:
add try catch

Commit:
Changing the ib yield filter to network config;
Fixing default apr handler error;
testing on all networks;

franzns committed:
ibyield flag

franzns committed:
cleanup

franzns committed:
dont use nested query for apr

timjrobinson commented:
I've been testing this out with https://github.com/balancer/frontend-v2/pull/4068 running it all locally, and found a few issues:

- All pools are missing BAL rewards

### Base

- USDbC rewards on cbETH/WETH (`0xfb4c2e6e6e27b5b4a07a36360c89ede29bb3c9b6000000000000000000000026`) and USDbC/axlUSDC (`0xe58ca65f418d4121d6c70d4c133e60cf6fda363c000000000000000000000013`) pools on base are way too low (like orders of magnitude lower than expected, seems like a scaling issue maybe because the reward token has 6 decimals rather than 18)
- Pool liquidity values are incorrect (different to prod)
- GOLD rewards APR are much lower than prod (40% vs 60%)
- There's no APR on `0x18f150c43598cd822d39711c55bd90407a8b8ad700020000000000000000003b` when in prod it has 12% APR - rewards from the subpool `0x2db50a0e0310723ef0c2a165cb9a9f80d772ba2f00020000000000000000000d`

### Ethereum

- `0x1e19cf2d73a72ef1332c882f20534b6519be0276000200000000000000000112` is missing rETH APR while `0x42ed016f826165c2e5976fe5bc3df540c5ad0af700000000000000000000058b` has the token APR's including rETH

For some things (like missing BAL rewards) I'm thinking maybe I haven't run the right jobs to get the data needed?

It would be good to get this merged to canary or hosted somewhere where it can continually update with the latest data and  anyone compare current Balancer prod with pulling pools from the beets API and report any differences.

franzns commented:
Thanks for testing further! We are in the final phase to test everything locally. Next phase is to merge to canary and have it running there to investigate differences between prod and canary, like you already did.

The missing/wrong BAL rewards is something we already noticed and @gmbronco is looking into that separately.

franzns committed:
Merge branch 'v3-canary' into non-reaper-yield-tokens-apr

franzns committed:
lower case all addresses in configs

franzns committed:
Merge branch 'v3-canary' into non-reaper-yield-tokens-apr

lgahdl commented:
> * 0x1e19cf2d73a72ef1332c882f20534b6519be0276000200000000000000000112
This rETH token included in the APR's is the stafi rETH (0x9559aaa82d9649c7a7b220e7c461d2e74c9a3593), not the rETH you're referring to (0xae78736cd615f374d3085123a210448e74fc6393), that's why this pool is not getting the rETH APR, thanks for seeing that.
 I'm not sure if the source we're using for the stafi rETH could be used for the rETH, but I can add it

lgahdl commented:
> I've been testing this out with [balancer/frontend-v2#4068](https://github.com/balancer/frontend-v2/pull/4068) running it all locally, and found a few issues:
> 
> * All pools are missing BAL rewards
> 
> ### Base
> * USDbC rewards on cbETH/WETH (`0xfb4c2e6e6e27b5b4a07a36360c89ede29bb3c9b6000000000000000000000026`) and USDbC/axlUSDC (`0xe58ca65f418d4121d6c70d4c133e60cf6fda363c000000000000000000000013`) pools on base are way too low (like orders of magnitude lower than expected, seems like a scaling issue maybe because the reward token has 6 decimals rather than 18)
> * Pool liquidity values are incorrect (different to prod)
> * GOLD rewards APR are much lower than prod (40% vs 60%)
> * There's no APR on `0x18f150c43598cd822d39711c55bd90407a8b8ad700020000000000000000003b` when in prod it has 12% APR - rewards from the subpool `0x2db50a0e0310723ef0c2a165cb9a9f80d772ba2f00020000000000000000000d`
> 
> ### Ethereum
> * `0x1e19cf2d73a72ef1332c882f20534b6519be0276000200000000000000000112` is missing rETH APR while `0x42ed016f826165c2e5976fe5bc3df540c5ad0af700000000000000000000058b` has the token APR's including rETH
> 
> For some things (like missing BAL rewards) I'm thinking maybe I haven't run the right jobs to get the data needed?
> 
> It would be good to get this merged to canary or hosted somewhere where it can continually update with the latest data and anyone compare current Balancer prod with pulling pools from the beets API and report any differences.

This rETH token included in the APR's is the stafi rETH (0x9559aaa82d9649c7a7b220e7c461d2e74c9a3593), not the rETH you're referring to (0xae78736cd615f374d3085123a210448e74fc6393), that's why this pool(0x1e19cf2d73a72ef1332c882f20534b6519be0276000200000000000000000112) is not getting the rETH APR, thanks for seeing that.
 I'm not sure if the source we're using for the stafi rETH could be used for the rETH, but I can add it
 Note.: The rETH APR of the pool 0x42ed016f826165c2e5976fe5bc3df540c5ad0af700000000000000000000058b is probably not from this service

franzns committed:
adding rocket eth apr
